File: 백준/Gold/17404. RGB거리 2/RGB거리 2.java
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int[][] P = new int[N+1][3];
		int[][] D = new int[N+1][3];
		for(int i=1; i<=N; i++) {
			String[] str = br.readLine().split(" ");
			for(int j=0; j<3; j++)
				P[i][j] = Integer.parseInt(str[j]);
		}
		int min = 1000*1000+1;
		for(int f=0; f<3; f++) {
			for(int c=0; c<3; c++)
				if(f==c)
					D[1][c] = P[1][c];
				else
					D[1][c] = 1000*1000+1;
			
			for(int i=2; i<=N; i++) {
				D[i][0] = Math.min(D[i-1][1], D[i-1][2]) + P[i][0];
				D[i][1] = Math.min(D[i-1][0], D[i-1][2]) + P[i][1];
				D[i][2] = Math.min(D[i-1][0], D[i-1][1]) + P[i][2];
			}
			
			for(int c=0; c<3; c++)
				if(f!=c)
					min = Math.min(min, D[N][c]);
		}
		System.out.print(min);
	}
}

File: 백준/Gold/17404. RGB거리 2/RGB거리 2.py
n = int(input())
cost = []
for i in range(n) :
    cost += [list(map(int, input().split()))]
ans = 1000*1000
for i in range(3) :
    d = [[0]*3 for _ in range(n)]
    for j in range(3) :
        if j == i :
            d[0][j] = cost[0][j]
        else :
            d[0][j] = 1000*1000
    for j in range(1, n) :
        d[j][0] = min(d[j-1][1], d[j-1][2]) + cost[j][0]
        d[j][1] = min(d[j-1][0], d[j-1][2]) + cost[j][1]
        d[j][2] = min(d[j-1][0], d[j-1][1]) + cost[j][2]
    for j in range(3) :
        if j != i :
            ans = min(ans, d[n-1][j])
print(ans)